### PR TITLE
[torchgen] Move Executorch unboxing logic into torchgen

### PR DESCRIPTION
**Pull request by larryliu0820**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89596
* #90806
* #90099
* __->__ #90098

This PR adds `unboxing.py` which converts a `EValue` (similar to `IValue`) to its corresponding C++ type, based on the `ExecutorchCppSignature`.

Added unit tests to it in `test_executorch_unboxing.py`. Notice that this unboxing logic should work for both ATen types and Executorch types, hence the unit tests are parametrized.

Differential Revision: [D42162582](https://our.internmc.facebook.com/intern/diff/D42162582)

### CODE REVIEW

File: torchgen/executorch/api/unboxing.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
+from dataclasses import dataclass
+from typing import Callable, List, Sequence, Tuple, Union
```
Comment (ezyang) [inline torchgen/executorch/api/unboxing.py:2]: What's the relationship between this and torchgen.api.unboxing? Can we delete torchgen.api.unboxing now?

File: torchgen/executorch/api/unboxing.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    Argument,
+    BaseTy,
+    BaseType,
+    ListType,
+    NativeFunction,
+    OptionalType,
+    Type,
+)
+
+connector = "\n\t"
```
Comment (ezyang) [inline torchgen/executorch/api/unboxing.py:17]: you sure you want to do this with tabs :think:

File: torchgen/executorch/api/unboxing.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    void mul_out(EValue** stack) {
+        EValue& self = *stack[0];
+        EValue& other = *stack[1];
+        EValue& out = *stack[2];
+        const torch::executor::Tensor & self_base = self.to<torch::executor::Tensor>();
+        const torch::executor::Tensor & other_base = other.to<torch::executor::Tensor>();
+        torch::executor::Tensor & out_base = out.to<torch::executor::Tensor>();
+
+        EXECUTORCH_SCOPE_PROF("native_call_mul.out");
+        torch::executor::mul_outf(self_base, other_base, out_base);
```
Comment (ezyang) [inline torchgen/executorch/api/unboxing.py:40]: This is not really the subject of the PR, but are you not responsible for popping the values off the stack?

File: torchgen/executorch/api/unboxing.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        self, arg_name: str, out_name: str, t: ListType, ctype: CType
+    ) -> Tuple[List[str], List[str]]:
+        in_name = f"{arg_name}_list_in"
+        elem_name = f"{arg_name}_elem"
+        code = []
+        res_name, res_ctype, res_code, decl = self.argumenttype_evalue_convert(
+            t.elem, elem_name
+        )
+
+        # pyre-fixme[16]: `Enum` has no attribute `Tensor`.
```
Comment (ezyang) [inline torchgen/executorch/api/unboxing.py:143]: How come pyre doesn't understand 🤔

File: torchgen/executorch/api/unboxing.py (hunk 1 above)
Comment (larryliu0820) [reply torchgen/executorch/api/unboxing.py:2]: no, `torchgen.api.unboxing` is doing unboxing for IValue. `torchgen.executorch.api.unboxing` is doing unboxing for EValue.

File: torchgen/executorch/api/unboxing.py (hunk 2 above)
Comment (larryliu0820) [reply torchgen/executorch/api/unboxing.py:17]: it doesn't really matter lol. what's the canonical way of doing this? whitespaces?

File: torchgen/executorch/api/unboxing.py (hunk 4 above)
Comment (larryliu0820) [reply torchgen/executorch/api/unboxing.py:143]: copied from fbcode, can remove.

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597383).

File: torchgen/executorch/api/unboxing.py (hunk 2 above)
Comment (ezyang) [reply torchgen/executorch/api/unboxing.py:17]: well, it's pretty tough to get f-strings to indent correctly, but yeah I usually just do spaces

**Review by ezyang — APPROVED** [review]
I am deeply suspicious that there should be a way to use translate here, but this is fine.

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597383).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42162582).

**Pull request closed without merging**